File: evaluations/multi-model-v1/stats.js
const round = (value) => Number(value.toFixed(6));
const mean = (values) => values.length ? values.reduce((sum, value) => sum + value, 0) / values.length : null;

export function wilsonInterval(successes, trials, z = 1.959964) {
  if (!trials) return { low: null, high: null };
  const p = successes / trials;
  const denominator = 1 + z ** 2 / trials;
  const center = (p + z ** 2 / (2 * trials)) / denominator;
  const margin = z * Math.sqrt((p * (1 - p) + z ** 2 / (4 * trials)) / trials) / denominator;
  return { low: round(Math.max(0, center - margin)), high: round(Math.min(1, center + margin)) };
}

function binomialTwoSided(successes, total) {
  if (!total) return 1;
  const choose = (n, k) => { let value = 1; for (let index = 1; index <= k; index += 1) value = value * (n - index + 1) / index; return value; };
  const tail = Array.from({ length: Math.min(successes, total - successes) + 1 }, (_, k) => choose(total, k) * 0.5 ** total).reduce((a, b) => a + b, 0);
  return Math.min(1, 2 * tail);
}

export function pairedGain(fromItems, toItems) {
  const from = new Map(fromItems.filter((item) => item.eligible).map((item) => [item.world_id, item]));
  const pairs = toItems.filter((item) => item.eligible && from.has(item.world_id)).map((item) => [from.get(item.world_id), item]);
  const differences = pairs.map(([a, b]) => Number(b.correct) - Number(a.correct));
  const gain = mean(differences) ?? 0;
  const variance = differences.length > 1 ? differences.reduce((sum, value) => sum + (value - gain) ** 2, 0) / (differences.length - 1) : 0;
  const margin = differences.length ? 1.959964 * Math.sqrt(variance / differences.length) : 0;
  const discordant = pairs.filter(([a, b]) => a.correct !== b.correct);
  const improvements = discordant.filter(([a, b]) => !a.correct && b.correct).length;
  return { gain: round(gain), pairs: pairs.length, confidence_interval_95: { low: round(Math.max(-1, gain - margin)), high: round(Math.min(1, gain + margin)) }, p_value: round(binomialTwoSided(improvements, discordant.length)) };
}

File: experiments/dri2/stats.py
"""Statistics for DRI-2, in the standard library only. DRAFT, NOT FROZEN."""

from __future__ import annotations

import math
from collections.abc import Mapping, Sequence
from fractions import Fraction

Z95 = 1.959963984540054


def wilson_interval(successes: int, trials: int, z: float = Z95) -> tuple[float | None, float | None]:
    if trials == 0:
        return None, None
    p = successes / trials
    z2 = z * z
    denominator = 1 + z2 / trials
    centre = (p + z2 / (2 * trials)) / denominator
    half = z * math.sqrt(p * (1 - p) / trials + z2 / (4 * trials * trials)) / denominator
    return max(0.0, centre - half), min(1.0, centre + half)


def mcnemar_exact(first_only: int, second_only: int) -> float:
    """Two-sided exact McNemar test on the discordant pairs."""
    n = first_only + second_only
    if n == 0:
        return 1.0
    tail = sum(math.comb(n, i) for i in range(min(first_only, second_only) + 1))
    return float(min(Fraction(1), Fraction(2 * tail, 2**n)))


def paired_difference_interval(
    first_only: int, second_only: int, pairs: int, z: float = Z95
) -> tuple[float | None, float | None, float | None]:
    """Difference in paired proportions (first minus second) with a Wald interval."""
    if pairs == 0:
        return None, None, None
    difference = (first_only - second_only) / pairs
    variance = (first_only + second_only - (first_only - second_only) ** 2 / pairs) / pairs**2
    half = z * math.sqrt(max(variance, 0.0))
    return difference, difference - half, difference + half


def holm(pvalues: Mapping[str, float]) -> dict[str, float]:
    """Holm step-down adjusted p-values."""
    ordered = sorted(pvalues.items(), key=lambda item: (item[1], item[0]))
    total = len(ordered)
    adjusted: dict[str, float] = {}
    running = 0.0
    for rank, (name, value) in enumerate(ordered):
        running = max(running, min(1.0, (total - rank) * value))
        adjusted[name] = running
    return adjusted


def wilcoxon_signed_rank(differences: Sequence[float]) -> dict[str, float | int]:
    """Two-sided Wilcoxon signed-rank test, normal approximation with tie and
    continuity correction. Zero differences are dropped."""
    nonzero = [value for value in differences if value != 0]
    n = len(nonzero)
    if n == 0:
        return {"n": 0, "wPlus": 0.0, "z": 0.0, "p": 1.0}
    order = sorted(range(n), key=lambda i: abs(nonzero[i]))
    ranks = [0.0] * n
    tie_sizes: list[int] = []
    start = 0
    while start < n:
        end = start
        while end + 1 < n and abs(nonzero[order[end + 1]]) == abs(nonzero[order[start]]):
            end += 1
        average = (start + end) / 2 + 1
        for position in range(start, end + 1):
            ranks[order[position]] = average
        if end > start:
            tie_sizes.append(end - start + 1)
        start = end + 1
    w_plus = sum(rank for rank, value in zip(ranks, nonzero) if value > 0)
    mean = n * (n + 1) / 4
    variance = n * (n + 1) * (2 * n + 1) / 24 - sum(t**3 - t for t in tie_sizes) / 48
    if variance <= 0:
        return {"n": n, "wPlus": w_plus, "z": 0.0, "p": 1.0}
    delta = w_plus - mean
    magnitude = max(abs(delta) - 0.5, 0.0) / math.sqrt(variance)
    z = math.copysign(magnitude, delta)
    return {"n": n, "wPlus": w_plus, "z": z, "p": min(1.0, math.erfc(magnitude / math.sqrt(2)))}


def mean_interval(values: Sequence[float], z: float = Z95) -> tuple[float | None, float | None, float | None]:
    n = len(values)
    if n == 0:
        return None, None, None
    mean = sum(values) / n
    if n < 2:
        return mean, None, None
    variance = sum((value - mean) ** 2 for value in values) / (n - 1)
    half = z * math.sqrt(variance / n)
    return mean, mean - half, mean + half
